N805: A wrong line number is reported when there is a multiline decorator

Consider the following sample code:

```python
class C:
    @decorator(
        'a')
    def m(cls):  # noqa: N805
        pass
````

run flake8 on it:

```
mytest.py:3:9: N805 first argument of a method should be named 'self'
```

The reported line number should be `mytest.py:4`.
This has also caused the `noqa` comment to become ineffective.

--- a/run_tests.py
+++ b/run_tests.py
@@ -13,7 +13,12 @@ IS_PY3 = sys.version_info[0] == 3
 IS_PY3_TEST = re.compile(r"^#\s*python3\s*only")
 IS_PY2_TEST = re.compile(r"^#\s*python2\s*only")
 
-TESTCASE_RE = re.compile("^#: (?P<code>\w+)(\((?P<options>.+)\))?$")
+TESTCASE_RE = re.compile(
+    r'#: '
+    r'(?P<code>\w+:?\d*:?\d*)'
+    r'(\((?P<options>.+)\))?'
+    r'$'
+)
 
 
 def main():
@@ -86,21 +91,24 @@ def parse_options(checker, options):
 
 
 def test_file(filename, lines, code, options):
+    if code is None:  # Invalid test case
+        return 0
     tree = compile(''.join(lines), '', 'exec', PyCF_ONLY_AST)
     checker = pep8ext_naming.NamingChecker(tree, filename)
     parse_options(checker, options)
+    error_format = (
+        '{0}:{lineno}:{col_offset}' if ':' in code else '{0}').format
 
-    found_errors = []
+    found_errors = set()
     for lineno, col_offset, msg, instance in checker.run():
-        found_errors.append(msg.split()[0])
+        found_errors.add(error_format(msg.split()[0], **locals()))
 
-    if code is None:  # Invalid test case
-        return 0
     if not found_errors and code == 'Okay':  # Expected PASS
         return 0
     if code in found_errors:  # Expected FAIL
         return 0
-    print("ERROR: %s not in %s" % (code, filename))
+    print("ERROR: %s not in %s. found_errors: %s"
+          % (code, filename, found_errors))
     return 1
 
 

--- a/src/pep8ext_naming.py
+++ b/src/pep8ext_naming.py
@@ -28,15 +28,18 @@ if PY2:
             if isinstance(arg, ast.Tuple):
                 ret.extend(_unpack_args(arg.elts))
             else:
-                ret.append(arg.id)
+                ret.append((arg, arg.id))
         return ret
 
-    def get_arg_names(node):
+    def get_arg_name_tuples(node):
         return _unpack_args(node.args.args)
+
+
 else:
-    def get_arg_names(node):
-        pos_args = [arg.arg for arg in node.args.args]
-        kw_only = [arg.arg for arg in node.args.kwonlyargs]
+    def get_arg_name_tuples(node):
+        args = node.args
+        pos_args = [(arg, arg.arg) for arg in args.args]
+        kw_only = [(arg, arg.arg) for arg in args.kwonlyargs]
         return pos_args + kw_only
 
 
@@ -59,7 +62,7 @@ def _err(self, node, code, **kwargs):
     code_str = getattr(self, code)
     if kwargs:
         code_str = code_str.format(**kwargs)
-    return (lineno, col_offset, '%s %s' % (code, code_str), self)
+    return lineno, col_offset + 1, '%s %s' % (code, code_str), self
 
 
 BaseASTCheck = _ASTCheckMeta('BaseASTCheck', (object,),
@@ -278,34 +281,33 @@ class FunctionArgNamesCheck(BaseASTCheck):
     def visit_functiondef(self, node, parents, ignore=None):
 
         def arg_name(arg):
-            return getattr(arg, 'arg', arg)
+            try:
+                return arg, arg.arg
+            except AttributeError:  # PY2
+                return node, arg
 
-        kwarg = arg_name(node.args.kwarg)
-        if kwarg is not None:
-            if not kwarg.islower():
-                yield self.err(node, 'N803', name=kwarg)
-                return
+        for arg, name in arg_name(node.args.vararg), arg_name(node.args.kwarg):
+            if name is not None:
+                if not name.islower():
+                    yield self.err(arg, 'N803', name=name)
+                    return
 
-        vararg = arg_name(node.args.vararg)
-        if vararg is not None:
-            if not vararg.islower():
-                yield self.err(node, 'N803', name=vararg)
-                return
 
-        arg_names = get_arg_names(node)
-        if not arg_names:
+        arg_name_tuples = get_arg_name_tuples(node)
+        if not arg_name_tuples:
             return
+        arg0, name0 = arg_name_tuples[0]
         function_type = getattr(node, 'function_type', 'function')
 
         if function_type == _FunctionType.METHOD:
-            if arg_names[0] != 'self':
-                yield self.err(node, 'N805')
+            if name0 != 'self':
+                yield self.err(arg0, 'N805')
         elif function_type == _FunctionType.CLASSMETHOD:
-            if arg_names[0] != 'cls':
-                yield self.err(node, 'N804')
-        for arg in arg_names:
-            if not arg.islower():
-                yield self.err(node, 'N803', name=arg)
+            if name0 != 'cls':
+                yield self.err(arg0, 'N804')
+        for arg, name in arg_name_tuples:
+            if not name.islower():
+                yield self.err(arg, 'N803', name=name)
                 return
 
 

--- a/testsuite/N801.py
+++ b/testsuite/N801.py
@@ -9,6 +9,6 @@ class Good(object):
 #: Okay
 class VeryGood(object):
     pass
-#: N801
+#: N801:1:7
 class _:
     pass

--- a/testsuite/N801_py3.py
+++ b/testsuite/N801_py3.py
@@ -11,7 +11,7 @@ class ΓγΓ6:
 #: Okay
 class _Γ:
     pass
-#: N801
+#: N801:1:7
 class γ:
     pass
 #: N801

--- a/testsuite/N802.py
+++ b/testsuite/N802.py
@@ -16,7 +16,7 @@ def go_od_():
 #: Okay
 def _go_od_():
     pass
-#: N802
+#: N802:1:5
 def NotOK():
     pass
 #: Okay

--- a/testsuite/N803.py
+++ b/testsuite/N803.py
@@ -37,10 +37,10 @@ def b12(*BAD):
 #: N803
 def b13(BAD, *VERYBAD, **EXTRABAD):
     pass
-#: N803
+#: N803:1:9
 def b14(BAD):
     pass
-#: N803
+#: N803:2:24
 class Test(object):
     def __init__(self, BAD):
         pass

--- a/testsuite/N804.py
+++ b/testsuite/N804.py
@@ -1,4 +1,4 @@
-#: N804
+#: N804:7:13
 class Foo(object):
     @classmethod
     def mmm(cls, ads):
@@ -14,7 +14,7 @@ class Foo(object):
 
     def __init_subclass(self, ads):
         pass
-#: N804(--classmethod-decorators=clazzy,cool)
+#: N804:3:14(--classmethod-decorators=clazzy,cool)
 class NewClassIsRequired(object):
     @cool
     def test(self, sy):

--- a/testsuite/N805.py
+++ b/testsuite/N805.py
@@ -1,3 +1,13 @@
+#: Okay
+class C:
+    def __init__(*args, **kwargs):
+        pass
+#: N805:4:11
+class C:
+    @decorator(
+        'a')
+    def m(cls, k='w'):  # noqa: N805
+        pass
 #: N805
 class Foo(object):
     def good(self, ads):

--- a/testsuite/N806.py
+++ b/testsuite/N806.py
@@ -104,7 +104,7 @@ def f():
 def f():
     for i in iterator:
         pass
-#: N806
+#: N806:2:9
 def f():
     for Bad in iterator:
         pass

--- a/testsuite/N807.py
+++ b/testsuite/N807.py
@@ -24,7 +24,7 @@ class C3:
 #: N807
 def __bad():
     pass
-#: N807
+#: N807:1:5
 def bad__():
     pass
 #: N807

--- a/testsuite/N815.py
+++ b/testsuite/N815.py
@@ -10,7 +10,7 @@ class C:
 #: Okay
 class C:
     _1 = 0
-#: N815
+#: N815:2:5
 class C:
     mixedCase = 0
 #: N815

--- a/testsuite/N816.py
+++ b/testsuite/N816.py
@@ -2,7 +2,7 @@
 GLOBAL_UPPER_CASE = 0
 #: N816
 mixedCase = 0
-#: N816
+#: N816:1:1
 mixed_Case = 0
 #: Okay
 _C = 0

--- a/testsuite/N81x.py
+++ b/testsuite/N81x.py
@@ -2,11 +2,11 @@
 import good as good
 #: Okay
 from mod import good as nice, NICE as GOOD, Camel as Memel
-#: N811
+#: N811:1:1
 from mod import GOOD as bad
-#: N812
+#: N812:1:1
 from mod import good as Bad
-#: N813
+#: N813:1:1
 from mod import CamelCase as noncamle
-#: N814
+#: N814:1:1
 from mod import CamelCase as CONSTANT

